cmd/snap-confine/seccomp-support.c

```diff
@@ -17,869 +17,182 @@
-	if (pre.complain) {
-		seccomp_release(ctx);
-		ctx = NULL;
-		goto out;
-	}
-
-	char buf[SC_MAX_LINE_LENGTH];
-	while (fgets(buf, sizeof(buf), f) != NULL) {
-		lineno++;
-
-		// skip policy-irrelevant lines
-		if (validate_and_trim_line(buf, sizeof(buf), lineno) == 0)
-			continue;
-
-		char *buf_copy = strdup(buf);
-		if (buf_copy == NULL)
-			die("Out of memory");
-
-		int pr_rc = parse_line(buf_copy, &sargs);
-		free(buf_copy);
-		if (pr_rc != PARSE_OK) {
-			// as this is a syscall whitelist an invalid syscall
-			// is ok and the error can be ignored
-			if (pr_rc == PARSE_INVALID_SYSCALL)
-				continue;
-			die("could not parse line");
-		}
-
-		rc = seccomp_rule_add_exact_array(ctx, SCMP_ACT_ALLOW,
-						  sargs.syscall_nr,
-						  sargs.length, sargs.arg_cmp);
-		if (rc != 0) {
-			rc = seccomp_rule_add_array(ctx, SCMP_ACT_ALLOW,
-						    sargs.syscall_nr,
-						    sargs.length,
-						    sargs.arg_cmp);
-			if (rc != 0) {
-				fprintf(stderr,
-					"seccomp_rule_add_array failed with %i for '%s'\n",
-					rc, buf);
-				errno = 0;
-				die("aborting");
-			}
-		}
-	}
-
- out:
-	if (f != NULL) {
-		if (fclose(f) != 0)
-			die("could not close seccomp file");
-	}
-	sc_map_destroy();
-	return ctx;
-}
-
-void sc_load_seccomp_context(scmp_filter_ctx ctx)
-{
-	int rc;
+	// validate '/' down to profile_path are root-owned and not
+	// 'other' writable to avoid possibility of privilege
+	// escalation via bpf program load when paths are incorrectly
+	// set on the system.
+	validate_bpfpath_is_safe(profile_path);
+
+	// sanity check size/owner
+	struct stat stat_buf;
+	if (stat(profile_path, &stat_buf) < 0)
+		die("cannot stat %s", profile_path);
+	if (stat_buf.st_size > MAX_BPF_SIZE)
+		die("profile %s is too big", profile_path);
+
+	// paranoid, but only helps a bit
+	if (stat_buf.st_uid != 0 || stat_buf.st_gid != 0)
+		die("cannot use %s: must be root:root owned", profile_path);
+
+	// load bpf
+	unsigned char bpf[MAX_BPF_SIZE + 1];	// acount for EOF
+	FILE *fp = fopen(profile_path, "rb");
+	if (fp == NULL)
+		die("cannot read %s", profile_path);
+	// set 'size' to 1 to get bytes transferred
+	size_t num_read = fread(bpf, 1, sizeof(bpf), fp);
+	if (ferror(fp) != 0)
+		die("cannot fread() %s", profile_path);
+	else if (feof(fp) == 0)
+		die("profile %s exceeds %zu bytes", profile_path, sizeof(bpf));
+	fclose(fp);
+	debug("read %zu bytes from %s", num_read, profile_path);
+
+	// validate bpf, it will die() if things look wrong
+	validate_bpf(bpf, num_read);
+
+	// raise privs
 	uid_t real_uid, effective_uid, saved_uid;
-
-	// if sc_prepare_seccomp_context() sees @unrestricted or @complain it bails
-	// out early and destroys the context object. In that case we have nothing
-	// to do.
-	if (ctx == NULL) {
-		return;
-	}
-
 	if (getresuid(&real_uid, &effective_uid, &saved_uid) != 0)
 		die("could not find user IDs");
-
 	// If not root but can raise, then raise privileges to load seccomp
 	// policy since we don't have nnp
```

Comment:
With the change to bpf loading, we have a behavioral change in that snap-confine will not be able to load policy in the kernel unless root, setuid root or has fscaps with CAP_SYS_ADMIN. We used to run cmd/snap-confine/tests/* as non-root during the build, but this will no longer work (that's ok for this PR, all those tests have been removed).

Also, this comment is a little misleading in the new world. Let's change it to be:
```
// If we can, raise privileges so that we can load the BPF into the
// kernel via 'prctl(PR_SET_SECCOMP, SECCOMP_MODE_FILTER, ...)'.
```